### PR TITLE
feat(whatsapp): add reaction guidance levels

### PR DESCRIPTION
**Pull request by mcaxtr**
## Summary

Describe the problem and fix in 2–5 bullets:

- Problem: WhatsApp exposed the `react` message action, but unlike Telegram and Signal it had no config-backed guidance/policy for when the agent should use reactions.
- Why it matters: reactions only happened when the agent was explicitly nudged, and operators had no way to distinguish ack-only behavior from agent-driven reactions.
- What changed: add `channels.whatsapp.reactionLevel` with Telegram/Signal-style semantics, default it to `extensive`, feed it into WhatsApp agent prompt guidance, and enforce it in the WhatsApp reaction action runtime.
- What did NOT change (scope boundary): no reply-payload/directive reaction delivery path was added; this PR only improves the existing `message(action="react")` behavior and policy.

## Change Type (select all)

- [ ] Bug fix
- [x] Feature
- [ ] Refactor required for the fix
- [x] Docs
- [ ] Security hardening
- [ ] Chore/infra

## Scope (select all touched areas)

- [ ] Gateway / orchestration
- [ ] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [x] Integrations
- [x] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #
- Related #
- [ ] This PR fixes a bug or regression

## Root Cause / Regression History (if applicable)

For bug fixes or regressions, explain why this happened, not just what changed. Otherwise write `N/A`. If the cause is unclear, write `Unknown`.

- Root cause: N/A
- Missing detection / guardrail: N/A
- Prior context (`git blame`, prior PR, issue, or refactor if known): Telegram and Signal already use reaction-level guidance and runtime gating; WhatsApp only had the raw action support.
- Why this regressed now: N/A
- If unknown, what was ruled out: N/A

## Regression Test Plan (if applicable)

For bug fixes or regressions, name the smallest reliable test coverage that should have caught this. Otherwise write `N/A`.

- Coverage level that should have caught this:
  - [ ] Unit test
  - [ ] Seam / integration test
  - [ ] End-to-end test
  - [x] Existing coverage already sufficient
- Target test or file: N/A
- Scenario the test should lock in: N/A
- Why this is the smallest reliable guardrail: N/A
- Existing test that already covers this (if any): new targeted tests were added in `extensions/whatsapp/src/reaction-level.test.ts`, `extensions/whatsapp/src/action-runtime.test.ts`, and `extensions/whatsapp/src/channel.test.ts`.
- If no new test is added, why not: N/A

## User-visible / Behavior Changes

- WhatsApp now supports `channels.whatsapp.reactionLevel: off | ack | minimal | extensive`.
- Default WhatsApp reaction guidance is `extensive` when reactions are enabled and no level is set.
- `off` and `ack` block agent-driven WhatsApp reactions while preserving ack-only behavior.
- Per-account override is supported via `channels.whatsapp.accounts.<id>.reactionLevel`.

## Diagram (if applicable)

N/A

## Security Impact (required)

- New permissions/capabilities? (`Yes/No`) No
- Secrets/tokens handling changed? (`Yes/No`) No
- New/changed network calls? (`Yes/No`) No
- Command/tool execution surface changed? (`Yes/No`) No
- Data access scope changed? (`Yes/No`) No
- If any `Yes`, explain risk + mitigation:

### Environment

- OS: Linux
- Runtime/container: Node 24 / pnpm / Bun
- Model/provider: N/A
- Integration/channel (if any): WhatsApp
- Relevant config (redacted): `channels.whatsapp.actions.reactions=true`; `channels.whatsapp.reactionLevel` varied across unset / `ack` / `minimal`

### Steps

1. Configure WhatsApp with reactions enabled and no explicit `reactionLevel`.
2. Resolve WhatsApp reaction guidance / effective level from the channel runtime.
3. Set `channels.whatsapp.reactionLevel: ack` and invoke the WhatsApp `react` action.

### Expected

- Unset `reactionLevel` resolves to `extensive` guidance.
- `minimal` and `extensive` enable agent reactions.
- `off` and `ack` disable agent reactions with a clear error.

### Actual

- Direct runtime verification resolved default WhatsApp reaction level to `extensive`.
- `minimal` guidance resolved as expected.
- `ack` rejected agent reactions with an explicit config error.
- `pnpm check` passed in the feature worktree.

## Evidence

Attach at least one:

- [ ] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

## Human Verification (required)

What you personally verified (not just CI), and how:

- Verified scenarios: `pnpm check`; direct Bun verification of `resolveWhatsAppReactionLevel` resolving `extensive` by default and `minimal` when configured; direct Bun verification that `handleWhatsAppAction(... action="react" ...)` rejects when `reactionLevel="ack"`.
- Edge cases checked: default-level fallback and ack-only runtime denial path.
- What you did **not** verify: targeted Vitest wrapper runs for the three WhatsApp tests did not produce a clean local completion in this worktree, so I did not get a stable passing wrapper result for those tests; no live WhatsApp session was exercised.

## Review Conversations

- [ ] I replied to or resolved every bot review conversation I addressed in this PR.
- [ ] I left unresolved only the conversations that still need reviewer or maintainer judgment.

If a bot review conversation is addressed by this PR, resolve that conversation yourself. Do not leave bot review conversation cleanup for maintainers.

## Compatibility / Migration

- Backward compatible? (`Yes/No`) Yes
- Config/env changes? (`Yes/No`) Yes
- Migration needed? (`Yes/No`) No
- If yes, exact upgrade steps:

## Risks and Mitigations

List only real risks for this PR. Add/remove entries as needed. If none, write `None`.

- Risk: `extensive` as the default can make WhatsApp agents react more often than before once reactions are enabled.
  - Mitigation: operators can set `reactionLevel` to `minimal`, `ack`, or `off`, and the action gate still applies.
- Risk: users relying on manual agent reactions while configured for `ack` will now get an explicit refusal.
  - Mitigation: the runtime error tells them to switch to `minimal` or `extensive`.

### CODE REVIEW

File: extensions/whatsapp/src/action-runtime.ts [hunk 1]
```diff
@@ -22,6 +23,17 @@ export async function handleWhatsAppAction(
...
+    const accountId = readStringParam(params, "accountId");
+    const reactionLevelInfo = resolveWhatsAppReactionLevel({
+      cfg,
+      accountId: accountId ?? undefined,
+    });
+    if (!reactionLevelInfo.agentReactionsEnabled) {
+      throw new Error(
+        `WhatsApp agent reactions disabled (reactionLevel="${reactionLevelInfo.level}"). ` +
+          `Set channels.whatsapp.reactionLevel to "minimal" or "extensive" to enable.`,
+      );
```
[the remark this answers is not in the document]
Comment (mcaxtr) [reply extensions/whatsapp/src/action-runtime.ts:42]: Aligned in b940cdf234: WhatsApp message-tool discovery now only advertises react when the effective reactionLevel allows agent reactions for the active/default account, and channel tests now cover the default, ack-disabled, and account-override discovery cases.

File: extensions/whatsapp/src/reaction-level.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  cfg: OpenClawConfig;
+  accountId?: string;
+}): ResolvedWhatsAppReactionLevel {
+  const account = resolveWhatsAppAccount({
+    cfg: params.cfg,
+    accountId: params.accountId,
+  });
+  return resolveReactionLevel({
+    value: account.reactionLevel,
+    defaultLevel: "minimal",
```
[the remark this answers is not in the document]
Comment (mcaxtr) [reply extensions/whatsapp/src/reaction-level.ts:23]: Aligned in 725ba59482. The documented default now matches the resolver behavior: `src/config/types.whatsapp.ts` says `minimal` is the default, and `extensions/whatsapp/src/reaction-level.ts` no longer duplicates stale per-level default prose.

File: src/config/zod-schema.providers-whatsapp.ts [hunk 3]
```diff
@@ -57,6 +57,7 @@ const WhatsAppSharedSchema = z.object({
   blockStreamingCoalesce: BlockStreamingCoalesceSchema.optional(),
   groups: WhatsAppGroupsSchema,
   ackReaction: WhatsAppAckReactionSchema,
+  reactionLevel: z.enum(["off", "ack", "minimal", "extensive"]).optional(),
```
[the remark this answers is not in the document]
Comment (mcaxtr) [reply src/config/zod-schema.providers-whatsapp.ts:60]: Addressed in 02ff665e1d. The branch now includes the regenerated config baselines, and both `docs/.generated/config-baseline.json` and `docs/.generated/config-baseline.jsonl` contain `channels.whatsapp.reactionLevel` plus `channels.whatsapp.accounts.*.reactionLevel`.

[the remark this answers is not in the document]
Comment (mcaxtr) [reply src/config/zod-schema.providers-whatsapp.ts:60]: Addressed in 02ff665e1d. The branch now includes the regenerated bundled channel metadata, and `src/config/bundled-channel-config-metadata.generated.ts` contains the WhatsApp `reactionLevel` entries at both the top level and under `accounts.*`.
